Zend/zend_opcode.c

```diff
@@ -1130,6 +1122,11 @@ ZEND_API void pass_two(zend_op_array *op_array)
 			case ZEND_JMP_FRAMELESS:
 				ZEND_PASS_TWO_UPDATE_JMP_TARGET(op_array, opline, opline->op2);
 				break;
+			case ZEND_FRAMELESS_ICALL_3:
+				// Copy result to OP_DATA to ensure dispatch in observer fallback can be aligned with the last opcode of the call
+				opline[1].result.var = opline->result.var;
+				opline[1].result_type = opline->result_type;
+				break;
```

Comment:
I didn't understand why this is necessary. For me this looks like a hack.

Reply:
For the fallback to ZEND_DO_ICALL (when we VM_DISPATCH) to work, the opcode after `opline` must be the next actual opline after and not OP_DATA. Hence ZEND_DO_ICALL must be dispatched on the OP_DATA opcode for ZEND_FRAMELESS_ICALL_3. And thus result must be on the OP_DATA opcode as well.